### ISSUE TITLE
'Cannot access Blockly on this page' Error causing addons not to load

### ISSUE THREAD

**Issue by Tacodude7729** [ScratchAddons/ScratchAddons#4212]
Sometimes, when switching from the project page to the editor, many addons will fail to work properly and the error below will be spammed in the console. I have seen this happen on at least three different occasions but, am unable to reliably reproduce it.

![image](https://user-images.githubusercontent.com/27910867/152663221-adf9df44-c3ed-4461-a1bb-ac530fcdbe08.png)

Firefox 96.0 64-bit
Ubuntu 20.04.3 (LTS)

**Comment by Samq64** [comment]
Is this the same bug as #4050?

**Comment by Tacodude7729** [comment]
I don't have either of those extensions enabled. I doubt this is an extension problem (it seems more like a core issue) but here is my extension list https://pastebin.com/BbYf5CFR

**Comment by WorldLanguages** [comment]
PR #4213 should provide more detailed logs.

**Comment by Tacodude7729** [comment]
Ok, I've pulled that down and I'll report back here if I see the error again.

**Comment by Chiroyce1** [comment]
> Ok, I've pulled that down and I'll report back here if I see the error again.

Can replicate, I updated my SA to v1.24.2 and it still occurs. Not at all times, but rarely.

<img width="696" alt="image" src="https://user-images.githubusercontent.com/97374054/153027516-77f47c42-aa38-4ae8-af71-e8de1675e5a7.png">

No devtools load, like Developer Tools, Debugger, pause button, etc.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/97374054/153027722-968ebf76-29bb-4822-944d-c24105682766.png">

**Comment by apple502j** [comment]
Okay, thanks for info! This issue is probably at https://github.com/ScratchAddons/ScratchAddons/blob/master/addon-api/content-script/Tab.js#L26 which runs too early. We should convert this to a more reliable solution - or a cached getter at the very least.

**Comment by WorldLanguages** [comment]
Yeah, if `addon.tab.traps._getEditorMode()` returns `false`, it can only mean that `addon.tab.traps._isWWW` is `false`. It should be `true` in project pages as long as the `<head>` has loaded completely.

I think this would be fixed if `_isWWW` was a function instead of checking once in the constructor. It's probably checking before the `<head>` loaded. Addons cannot run before the head loads, so I don't find it necessary to change how scratch-www is detected.

```js
// Traps.js right now:
export default class Trap extends Listenable {
  constructor(tab) {
    super();
    this._react_internal_key = undefined;
    this._isWWW = tab.clientVersion === "scratch-www"; // ← ← ← ← ← ← ← ← ← ← ← ← ← HERE
    this._getEditorMode = () => this._isWWW && tab.editorMode;
    this._waitForElement = tab.waitForElement.bind(tab);
    this._cache = Object.create(null);
  }
// ...
```

It should be changed to:
```js
this._isWWW = () => tab.clientVersion === "scratch-www";
this._getEditorMode = () => this._isWWW() && tab.editorMode;
```

`_getEditorMode` runs when an addon requests the vm or blockly, and [addons cannot run before the head loads](https://github.com/ScratchAddons/ScratchAddons/blob/1f94c03e96cfc4c6a7c76f0a6adf24280082ad6c/content-scripts/inject/module.js#L210), so this would fix the issue.

**Comment by WorldLanguages** [comment]
We're quickly getting into browser engineering and specs territory, lol.

Here's what I could observe:
1. `document.head` might be `null` in content scripts if they run at `document_start`. I believe it's null when the bytes that represent the string `<head>` haven't arrived yet.
2. `document.head` being truthy does not mean the head has loaded completely! This is specially noticeable if the head has many nodes inside it.
3. `<script type="module">` scripts usually run just before `DOMContentLoaded`. However, our `module.js` file is not part of the source, it's added dynamically, so it usually runs before that.
4. As far as I tested, `document.head` is always truthy inside `module.js`, but it's probably not guaranteed to be. I could easily make `document.body` return null, unlike typical modules that aren't added dynamically, where it was always truthy (makes sense, because those run just before `DOMContentLoaded`)
5. I'm not sure why we run `runAtComplete: true` addons on `load` instead of `DOMContentLoaded` 🤔

**Comment by WorldLanguages** [comment]
It's not the perfect solution, but we're launching v1.24.3 already, so I'll convert `_isWWW` to a function, which should fix the issue.

The fact `document.head` being truthy does not mean that the head loaded completely may cause some race conditions in certain situations too, but let's keep that for a future release.

**Comment by WorldLanguages** [comment]
Whoops, I forgot `addon.tab.clientVersion` is not a function either! So I'll need to change that too. `clientVersion` will only be calculated the first time an addon requests it.

**Comment by WorldLanguages** [comment]
Wow, it's hard to wrap my head around this.

So, currently on the stable release, `clientVersion` is calculated independently for each addon when `new Trap()` is called by `new Tab()` which is by itself called by `new Addon()` on content-scripts/inject/run-userscript.js.
The function on that module is called by the `runUserscripts` function on module.js, which itself is called when `document.head` is truthy:

```js
  const runUserscripts = () => {
    for (const addon of addons) {
      if (addon.scripts.length) runAddonUserscripts(addon);
    }
  };

  // Note: we currently load userscripts and locales after head loaded
  // We could do that before head loaded just fine, as long as we don't
  // actually *run* the addons before document.head is defined.
  if (document.head) runUserscripts();
  else {
    const observer = new MutationObserver(() => {
      if (document.head) {
        runUserscripts();
        observer.disconnect();
      }
    });
    observer.observe(document.documentElement, { subtree: true, childList: true });
  }
}
```

So this all means that sometimes, `document.head` is truthy even tho the DOM hasn't loaded the `<meta name="format-detection" content="telephone=no">` element yet. In other words, `</head>` wasn't written to the DOM, but `<head>` was, which confirms what I just said an hour ago.
This is a pretty amazing race condition because network conditions affect this, hehe TCP!

**Comment by WorldLanguages** [comment]
So taking that into account, it looks like a fix like #4244 would only mitigate this, not fix completely. Userscripts with `"runAtComplete": false` could theoretically still run `addon.tab.clientVersion` before the head loaded completely. However, locales and JavaScript needs to be fetched before the addon can request clientVersion, making it less likely.
It would also affect less addons, since only a minority has runAtComplete set to false. It previously affected all addons because `new Addon()` calls for every addon happen in the same event loop cycle, and clientVersion used to be calculated at that point, but that wouldn't be true anymore if the fix is merged.

**Comment by apple502j** [comment]
Potential solution is to use the same length regex in background to check for WWW-ness. That way we only need pathname. Hacky though.

**Comment by WorldLanguages** [comment]
We should be guaranteeing that the `<head>` is fully loaded before running addons and adding styles, so I don't think it's necessary to use a regex. If we can guarantee the head has fully loaded, we can keep the same clientVersion logic we've always used.

**Comment by RedGuy12** [comment]
> Potential solution is to use the same lengthy regex in background to check for WWW-ness. That way we only need pathname. Hacky though.

I dislike the regex due to its hockey nature

**Comment by WorldLanguages** [comment]
> we could use the DOM logic for matching

@RedGuy12 That's what we have always done and will continue to do, but we'll be more careful now, and try to consider edge cases.

**Comment by RedGuy12** [comment]
> > we could use the DOM logic for matching
> 
> @RedGuy12 That's what we have always done and will continue to do, but we'll be more careful now, and try to consider edge cases.

...no, we use a regex?

**Comment by mxmou** [comment]
> > we could use the DOM logic for matching
> 
> @RedGuy12 That's what we have always done and will continue to do, but we'll be more careful now, and try to consider edge cases.

I think @RedGuy12 is talking about userscript matches. The DOM can't be used for that because it's done in the background script.

**Comment by WorldLanguages** [comment]
I need to do more testing, but I'm pretty sure that if `document.body` is truthy (not null), then the head is guaranteed to have loaded completely.

**Comment by WorldLanguages** [comment]
See #5962 for a proposed fix for this issue
